In [2]:
# Import the modules
import numpy as np
import pandas as pd
import hvplot.pandas
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [6]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv(
    Path("Resources/lending_data.csv")
)

# Review the DataFrame
lending_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0          10700          7.672            52800        0.431818   
1           8400          6.692            43600        0.311927   
2           9000          6.963            46100        0.349241   
3          10700          7.664            52700        0.430740   
4          10800          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531      19100         11.261            86600        0.653580   
77532      17700         10.662            80900        0.629172   
77533      17600         10.595            80300        0.626401   
77534      16300         10.068            75300        0.601594   
77535      15600          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [7]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_df['loan_status']

# Separate the X variable, the features
X = lending_df.drop(columns=['loan_status'])

In [8]:
# Review the y variable Series
y

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64

In [9]:
# Review the X variable DataFrame
X

loan_size  interest_rate  borrower_income  debt_to_income  \
0          10700          7.672            52800        0.431818   
1           8400          6.692            43600        0.311927   
2           9000          6.963            46100        0.349241   
3          10700          7.664            52700        0.430740   
4          10800          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531      19100         11.261            86600        0.653580   
77532      17700         10.662            80900        0.629172   
77533      17600         10.595            80300        0.626401   
77534      16300         10.068            75300        0.601594   
77535      15600          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [11]:
# Check the balance of our target values
y_counts = y.value_counts()
y_counts

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [12]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [13]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier = LogisticRegression(solver='lbfgs', random_state=1)
classifier

# Fit the model using training data
classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

In [14]:
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.9914878250103177
Testing Data Score: 0.9924164259182832


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [17]:
# Make a prediction using the testing data
predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test}).tail(10)

Prediction  Actual
76038           1       1
19119           0       0
45193           0       0
25211           0       0
61697           0       0
38069           0       0
36892           0       0
5035            0       0
40821           0       0
35030           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import accuracy_score
accuracy_score(y_test, predictions)

0.9924164259182832

In [18]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, predictions)

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [20]:
# Print the classification report for the model
from sklearn.metrics import classification_report
target_names = ["Healthy Loan", "High-Risk Loan"]
print(classification_report(y_test, predictions, target_names=target_names))

                precision    recall  f1-score   support

  Healthy Loan       1.00      1.00      1.00     18759
High-Risk Loan       0.87      0.89      0.88       625

      accuracy                           0.99     19384
     macro avg       0.94      0.94      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The model works wonderfully for healthy loans, but less so for high-risk loans.  Healthy loans show 100% accuracy.  High-risk loans show an average of 88% accuracy.  Overall accuracy is good, however, macro average which looks at precision, recall, and f1-score for both healthy and at-risk loans is lower at 94%.  Weighted average looks at the class sizes and the number of samples in each class and is nearly equal to the healthy loan numbers.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [23]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
random_over_sampler = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = random_over_sampler.fit_resample(X_train, y_train)

In [24]:
# Count the distinct values of the resampled labels data
y_resampled_counts = y_resampled.value_counts()
y_resampled_counts

loan_status
0    56277
1    56277
Name: count, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [25]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
classifier_resampled = LogisticRegression(solver='lbfgs', random_state=1)
classifier_resampled

# Fit the model using the resampled training data
classifier_resampled.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
predictions_resampled = classifier_resampled.predict(X_test)
pd.DataFrame({"Prediction": predictions_resampled, "Actual": y_test}).tail(10)

Prediction  Actual
76038           1       1
19119           0       0
45193           0       0
25211           0       0
61697           0       0
38069           0       0
36892           0       0
5035            0       0
40821           0       0
35030           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [26]:
# Print the balanced_accuracy score of the model 
accuracy_score(y_test, predictions_resampled)

0.9952022286421791

In [27]:
# Generate a confusion matrix for the model
confusion_matrix(y_test, predictions_resampled)

array([[18668,    91],
       [    2,   623]], dtype=int64)

In [28]:
# Print the classification report for the model
target_names_resampled = ["Healthy Loan", "High-Risk Loan"]
print(classification_report(y_test, predictions_resampled, target_names=target_names_resampled))

                precision    recall  f1-score   support

  Healthy Loan       1.00      1.00      1.00     18759
High-Risk Loan       0.87      1.00      0.93       625

      accuracy                           1.00     19384
     macro avg       0.94      1.00      0.96     19384
  weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Interestingly, the accuracy for high-risk loans greatly improved in some areas with the oversampling.  The accuracy score improved slightly from .9924 to .9952.  The confusion matrix shows a slight decrease in accuracy for healthy loans from 80 false positives to 91, but high-risk loans improved from 67 false positives to 2!  

The classification report is interesting, in that there was no change for precision, but recall improved from .89 for high-risk loans to 1.00, and f1-score improved from .88 to .93.  For accuracy in the the classification report, precision showed no change, but recall improved from .94 macro/.99 weighted to 1.00 / 1.00 and f1-score improved from .99 accuracy / .94 macro / .99 weighted to 1.00 / .96 / 1.00.

It is clear that creating more data for training the model results in better accuracy for the test data.